README.txt

    Edward C. Eberle <eberdeed@eberdeed.net>
    San Diego, California USA
    
    sdl2fireworks is a program that displays a fireworks
    display with sound.  The keys are as follows.
    a right
    s back
    d left
    w forward
    r up
    f down
    x reverse view.
    z reset view.
    Escape ends the program.
    Alt+Return sets full screen.
    Up arrow zooms in.
    Down arrow zooms out.
    
    To compile the program you will need the following libraries:
    FreeImage, GLEW, SDL2, boost, pthread and GLM. 
    You must also have cmake.
    The commands are:
    
    cd build
    cmake ..
    sudo make
    sudo make doc
    sudo make install
    
    To run the program:
    
    sdl2fireworks
    
    The documentation is located in:
    
    /usr/share/doc/sdl2fireworks-doc
    
    The sources I used to educate myself concerning OpenGL are
    as follows:
        "OpenGL ES 3.0 Programming Guide Second Edition" 
    by Dan Ginsburg and Budirijanto Purnomo published by Addison-Wesley 2014.
        www.learnopengl.com.
    GLEW http://glew.sourceforge.net, and
    CMAKE www.cmake.org.
    SDL2 https://wiki.libsdl.org/FrontPage
    pthread https://www.gnu.org/software/hurd/libpthread.html
    OpenGL is generally available on any system as part 
    of the underlying graphics subsystem and GLEW will 
    get you access to it.
    
    Documentation is available after install at:
    
    /usr/share/doc/sdl2fireworks-doc
    
        
    The program was written and compiled on debian linux available at:  www.debian.org.
    
    You can reach me at <eberdeed@eberdeed.net>.
    Updates will be available at www.eberdeed.net.
    Edward Charles Eberle 
    April 17th, 2020 San Diego, California United States of America
    
    This program is licensed under the Lesser GNU Public license.
    It can be found in the assimpopengl/LGPL directory 
    and once installed it will be also found at 
    /usr/share/doc/sdl2fireworks-doc/LGPL.  This 
    program is given as an instructional aid for learning
    and using the libraries presented and is not to 
    be considered fit for any particular use.
    
    


CHANGELOG.txt

    Change Log for SDL2Fireworks

    03/24/20 Initial program written and documented.
    ECE

    04/18/20 Documented and removed the Framebuffer object.
    ECE


include/camera.h
/*******************************************************************
 * Camera:  A class to contain the functions of an OpenGL camera.
 *  Adapted from a class proposed on www.learnopengl.com.
 * Edward C. Eberle <eberdeed@eberdeed.net>
 * April 2020 San Diego, California USA
 * ****************************************************************/

#ifndef CAMERA_H
#define CAMERA_H

#include "commonheader.h"

/** \class Camera
 * A camera class that processes input and calculates the 
 * corresponding Euler Angles, Vectors and Matrices for moving the 
 * camera about a 3-dimensional landscape. For use in OpenGL.  This 
 * class relies heavily on GLM the OpenGL Math Library.
 */
class Camera
{
public:

    /** \brief Defines several possible options for camera movement. Used as 
     * abstraction to stay away from window-system specific input methods.
     * To access this from the calling class use 
     * "Camera::Camera_Movement::FORWARD, etc.
     */
    enum Camera_Movement {
        FORWARD,
        BACKWARD,
        LEFT,
        RIGHT,
        CLOSER,
        AWAY,
        UP,
        DOWN
    };
    //! Default camera values
    constexpr static float YAW         = -90.0f;
    constexpr static float PITCH       =  0.0f;
    const float SPEED       =  0.1f;
    const float SENSITIVITY =  0.005f;
    const float ZOOM        =  45.0f;
    const vec3 WORLDUP      =  vec3(0.0f, -1.0f, 0.0f);
    const float onedegree = acos(-1.0f) / 180.0f;
    const float pi90 = acos(-1.0f) / 2.0f;
    const float pi180 = acos(-1.0f);
    const float pi360 = acos(-1.0f) * 2.0f;
    
    //! Camera Attributes
    vec3 Position, Front, Up, Focus, Right, WorldUp;
    //! Persistent camera attributes.
    vec3 position, focus, front;
    //! Euler Angles current and persistent.
    float Yaw, yaw, Pitch, pitch;
    //! Camera options
    float MovementSpeed;
    float MouseSensitivity;
    //! Zoom for a more detailed view of an object.
    float Zoom;
    //! Viewport dimensions.
    int Width, Height, width, height;
    //! Perspective matrix.
    mat4 projection = mat4(1.0f);
    //! Debug flag.
    bool debug1 = false;
    /** \brief Constructor with vectors to define viewer position and focus.
     * width : the width of the viewport.
     * height : the height of the viewport.
     * position : the position of the camera.
     * focus :  the position of the object looked at.
     */
    Camera(int width, int height, vec3 position = vec3(0.0f, 0.0f, 2.0f), vec3 focus = vec3(0.0f, 0.0f, 0.0f)); 
    /** \brief Constructor with values to define viewer postion and Euler angles for direction.
     * width : the width of the viewport.
     * height : the height of the viewport.
     * posX, posY, posZ : the (x, y, z) position of the camera.
     * upX, upY, upZ : the (x, y, z) direction of camera up.
     * yaw : yaw based on (0, y, z > 0) as zero radians.
     * pitch : pitch based on (x, 0, z) as zero radians.
     */
    Camera(int width, int height, float posX, float posY, float posZ, float upX, float upY, 
           float upZ, float yaw, float pitch);
    /** \brief Destructor to echo the destruction of the class.
     */
    ~Camera();
    /** \brief Returns the LookAt Matrix.
     */
    mat4 getViewMatrix();
    /**  \brief Returns the current perspective matrix.
     */
    mat4 getPerspective();
    /** \brief Allows for viewport resize.
     */
    void resizeView(int width, int height);
    /** \brief Return the camera to its original settings.
     */
    void resetCamera();
    /** \brief Get the camera position.
     */
    vec3 getPosition();
    /** \brief Set the camera position.
     */
    void setPosition(vec3 position);
    /** \brief Rotate the camera 180 degrees
     *  on the XZ plane.
     */
    void reverseDirection();
    /** \brief Processes input received from any keyboard-like input system. 
     * Accepts input parameter in the form of camera defined ENUM 
     * (to abstract it from windowing systems). This is adjustable using
     * the SPEED variable.
     */
    void processKeyboard(Camera_Movement direction, float deltaTime);
    
    /** \brief Processes input received from a mouse input system. 
     * Expects the offset value in both the x and y direction.
     * This is adjustable using the SENSITIVITY variable.
     */
    void processMouseMovement(float xoffset, float yoffset);
    /** \brief Processes input received from a mouse scroll-wheel event. 
     * Only requires input on the vertical wheel-axis
     */
    void processMouseScroll(Camera_Movement inout);
    /** \brief Calculates the Euler angles from the front vector.
     */
    void getEulerAngles();
    /** \brief Calculate the forward vector from the Euler angles.
     */
    void getFront();
   
};

#endif // CAMERA_H


include/shader.h
/*******************************************************************
 * Shader:  A class to encapsulate the creation and use of a set of
 * shaders.  
 * Edward C. Eberle <eberdeed@eberdeed.net>
 * March 2020 San Diego, California USA
 * ****************************************************************/

#ifndef SHADER_H
#define SHADER_H

#include "commonheader.h"

using namespace std;
using namespace glm;
/** \class Shader A class to encapsulate the uploading, compiling, 
 * linking and use of a shader.  This class will create a shader 
 * binary and reload it.  The binary will be recreated if it ceases 
 * to validate as a shader program.
 */

class Shader
{
public:
    /** \brief The constructor signals the creation of the class.
     */
    Shader();
    /** \brief The destructor signals the destruction of the class,
     *  and deletes the shader program.
     */  
    ~Shader();
    /** \brief Read and build the shader program from two files,
     *  and provide the file name for the created binary.
     *  Delete the two shader objects afterwards.
     *  The variables are as follows:
     *  vertexPath : The vertex shader's location in the file structure.
     *  fragmentPath : The fragment shader's location in the file structure.
     *  outputFile : The name given to the output binary.
     */
    void initShader(string vertexPath, string fragmentPath, 
    string outputFile);
    /** \brief Create either the vertex or fragment shader from a file.
     */
    unsigned int createShader(unsigned int type, string fpath);
    /** \brief Use the program object for display.
     */
    void Use();
    /** \brief Create the shader program binary save it to a file,
     * and delete the binary pointer afterward.
     */
    bool createBinary();
    /** \brief Utility uniform functions that set values in the shader program.
     */
    void setBool(const string name, bool value) const;  
    void setInt(const string name, int value) const;   
    void setFloat(const string name, float value) const;
    void setVec2(const std::string name, vec2 value) const;
    void setVec3(const std::string name, vec3 value) const;
    void setVec4(const string name, vec4 value) const; 
    void setMat4(const string name, mat4 value) const;    
    
    /** Variables **/
    //! The shader program object.
    GLuint Program;
    //! Class global variables.
    //! Debugging information.
    bool debug1 = false;
    //! Indicates success of OpenGL function calls.
    int success = 0;
    //! Length of an OpenGL error information message.
    int infoLength = 0;
    //! Handles for vertex and fragment shaders respectively.
    GLuint vertex, fragment;
    //! The length of the program in bytes.
    int progLength = 0;
    //! The length of the program binary in bytes.
    int progLenRet = 0;
    //! Formats for the stored binary program.
    GLenum *valFormats;
    GLenum format = 0;
    //! A variable indicating the successful loading of a stored program binary.
    int response = 0;
    //! The binary program pointer.
    unsigned char *binary;
    /** The output file name.  The final stored binary is stored
     * int the .config directory of the user's home directory (~/.config).
     */
    string outputFile;

};
  
#endif //SHADER_H


include/commonheader.h
/*******************************************************************
 * CommonHeader:  A header file to reduce repetition of #define
 * and #include statements.
 * Edward C. Eberle <eberdeed@eberdeed.net>
 * April 2020 San Diego, California USA
****************************************************************/
#ifndef COMMONHEADER_H
#define COMMONHEADER_H
/** \class CommonHeader A header file to hold all the includes, defines and
 *  structs the program uses.
 */
//! The defines.
#define NUM_PARTICLES 10000
#define NUM_EXPLOSIONS 3
#define EXP_GAP 200
#define SCALE 15
//! GLEW The OpenGL library manager
#define GLEW_STATIC
#include <glew.h>
//! GLM The OpenGL math library
#define GLM_FORCE_RADIANS
#define GLM_ENABLE_EXPERIMENTAL
#include <glm.hpp>
#include <gtc/matrix_transform.hpp>
#include <gtx/quaternion.hpp>
#include <gtc/quaternion.hpp>
#include <gtc/type_ptr.hpp>
#include <gtx/matrix_interpolation.hpp>
#include <gtc/matrix_access.hpp> 
#include <gtx/euler_angles.hpp>
//! SDL2
#include <SDL.h>
//! CPP Library
#include <string>
#include <fstream>
#include <iostream>
#include <exception>
#include <cstdio>
#include <cstdlib>
#include <random>
#include <sstream>
#include <chrono>
#include <vector>
//! Boost
#include <filesystem.hpp>
//! System
#include <pthread.h>

using namespace std;
using namespace glm;
using namespace boost::filesystem;

//! The structures.
//! Define a Partilcle.
struct ParticleData {
    vec4 startPos;
    vec4 endPos;
    vec4 lifetime;
};
//! Define an explosion.
struct InitData {
    vec4 centerPos;
    vec4 color;
    unsigned int gap;
    bool start;
};

#endif // COMMONHEADER_H


src/sdl2fireworks.cpp
/**********************************************************
 *   SDL2Fireworks:  A class to encapsulate creating an 
 *   OpenGL window.
 *   Created by: Edward Charles Eberle <eberdeed@eberdeed.net>
 *   12/2019 San Diego, California USA
 * ********************************************************/

#include "../include/sdl2fireworks.h"


SDL2Fireworks::SDL2Fireworks()
{
    cout << "\n\n\tCreating SDL2Fireworks\n\n";
    quit = false;
    camera = new Camera(SCR_WIDTH, SCR_HEIGHT, vec3(0.0f, 0.0f, 18.0f), vec3(0.0f, 0.0f, 0.0f)); 
    tdata = chrono::system_clock::now();
    // Fill in particle data array
    srand ((long int)chrono::system_clock::to_time_t(tdata));
    // Create the data.
    initFire();
    scaler = mat4(1.0);
    scaler = scale(scaler, vec3((float)SCALE, (float)SCALE, (float)SCALE));
    width = SCR_WIDTH;
    height = SCR_HEIGHT;
    execLoop();
}

SDL2Fireworks::~SDL2Fireworks()
{
    cout << "\n\n\tDestroying SDL2Fireworks\n\n";
    delete camera;
    for (int x = 0; x < NUM_EXPLOSIONS; x++)
    {
        delete shader[x];
    }
    glDeleteVertexArrays(1, &VAO);
}

void SDL2Fireworks::initFire()
{
    bool isUnique = true;
    vec3 *pos;
    cout << "\n\n\tIn initSound.\n\n";
    for (int i = 0; i < NUM_PARTICLES; i++ )
    {
        do
        {
            pos = genCoords();
            // Lifetime of particle
            if (i == 0)
            {
                chckNorm[0] = pos[2];
            }
            else
            {
                isUnique = testNorm(pos[2], i);
            }
        } while(!isUnique);
        // Lifetime of particle
        particles[i].lifetime = vec4(0.0f, 0.0f, 0.0f, 0.0f);
        particles[i].lifetime.x = (float)(rand() % 500) + 500;
        particles[i].startPos = vec4(pos[0], 1.0f);
        particles[i].endPos = vec4(pos[1], 1.0f);
    }
    if (debug1)
    {
        debug();
    }
    timebegin = chrono::system_clock::now();
}

void SDL2Fireworks::debug()
{
        for (int x = 0; x < NUM_PARTICLES; x++)
    {
        cout << "\n\tStart:  " << particles[x].startPos.x 
        << ", " << particles[x].startPos.y << ", "
        << particles[x].startPos.z << ", " 
        << particles[x].startPos.w << "  End:  "
        << particles[x].endPos.x << ", " 
        << particles[x].endPos.y << ", " 
        << particles[x].endPos.z << ", " 
        << particles[x].endPos.w << "  Lifetime:  "
        << particles[x].lifetime.x;
    }
}
vec3* SDL2Fireworks::genCoords()
{
    vec3 *pos, startPos, endPos, normal;
    pos = new vec3[3];
    float startdist, enddist;
        // End position of particle
        if ((rand() % 1000) >= 500)
        {
            sign = -1.0f;
        }
        else
        {
            sign = 1.0f;
        }
        endPos.x = sign *((((float)(rand() % 5000) / 10000.0f) + 0.5f) - 1.0f) / 2.0f;
        if ((rand() % 1000) >= 500)
        {
            sign = -1.0f;
        }
        else
        {
            sign = 1.0f;
        }
        endPos.y = sign * ((((float)(rand() % 5000) / 10000.0f) + 0.5f) - 1.0f) / 2.0f;
        if ((rand() % 1000) >= 500)
        {
            sign = -1.0f;
        }
        else
        {
            sign = 1.0f;
        }
        endPos.z = sign * ((((float)(rand() % 5000) / 10000.0f) + 0.5f) - 1.0f) / 2.0f;
        // Start position of particle
        if ((rand() % 1000) >= 500)
        {
            sign = -1.0f;
        }
        else
        {
            sign = 1.0f;
        }
        startPos.x = sign * (((float)(rand() % 10000) / 40000.0f) - 0.125f) / 2.0f;
        if ((rand() % 1000) >= 500)
        {
            sign = -1.0f;
        }
        else
        {
            sign = 1.0f;
        }
        startPos.y = sign * (((float)(rand() % 10000) / 40000.0f) - 0.125f) / 2.0f;
        if ((rand() % 1000) >= 500)
        {
            sign = -1.0f;
        }
        else
        {
            sign = 1.0f;
        }
        startPos.z = sign * (((float)(rand() % 10000) / 40000.0f) - 0.125f) / 2.0f;
        normal = normalize(endPos - startPos);
        enddist = ((float)(rand() % 10000)) / 20000.0f;
        startdist = ((float)(rand() % 10000)) / 1000000.0f;
        pos[0] = normal * startdist;
        pos[1] = normal * enddist;
        pos[2] = normal;
        return(pos);
}

bool SDL2Fireworks::testNorm(vec3 normal, int x)
{
    for (int i = 0; i < x; i++)
    {
        if ((normal.x == chckNorm[i].x) && (normal.y == chckNorm[i].y) &&
        (normal.z == chckNorm[i].z))
        {
            return false;
        }
    }
    chckNorm[x] = normal;
    return true;
}

///
//  Update time-based variables
//
void SDL2Fireworks::update(int startup, int shaderVal)
{
    vec4 centerPos;
    vec4 color;
    if (debug2)
    {
        cout << "\n\n\tIn update for " << shaderVal << ".\n\n";
    }
    
    // Pick a new start location and color
    if ((rand() % 1000) >= 500)
    {
        sign = -1.0f;
    }
    else
    {
        sign = 1.0f;
    }
    centerPos.x = sign * (((float) (rand() % 10000) / 10000.0f) - 0.5f);
    if ((rand() % 1000) >= 500)
    {
        sign = -1.0f;
    }
    else
    {
        sign = 1.0f;
    }
    centerPos.y = sign * (((float) (rand() % 10000) / 10000.0f) - 0.5f);
    centerPos.z = -(((float) (rand() % 10000) / 10000.0f) - 0.5f);
    centerPos.w = 1.0f;
    // Random color
    color.r = ((float) (rand() % 60000) / 100000.0f) + 0.3f;
    color.g = ((float) (rand() % 60000) / 100000.0f) + 0.3f;
    color.b = ((float) (rand() % 60000) / 100000.0f) + 0.3f;
    color.a = 0.5f;
    if (debug1)
    {
        cout << "\n\n\tColor: " << color.r << ", " << color.g << ", " 
        << color.b << ", " <<  color.a;
    }
    explData[shaderVal].gap = startup + 500 + rand() % 500;
    for (int x = 0; x < NUM_EXPLOSIONS; x++)
    {
        if ((shaderVal != x) && (explData[shaderVal].gap > (explData[x].gap - EXP_GAP)) 
        && (explData[shaderVal].gap < (explData[x].gap + EXP_GAP)))
        {
            explData[shaderVal].gap = startup + 500 + rand() % 500;
            x = 0;
        }
    }
    explData[shaderVal].centerPos = centerPos;
    explData[shaderVal].color = color;
    explData[shaderVal].start = true;
    shader[shaderVal]->setFloat("u_time", 0.0f);
    shader[shaderVal]->setVec3("u_centerPosition", explData[shaderVal].centerPos);
    shader[shaderVal]->setVec4("u_color", explData[shaderVal].color);
    
}

void SDL2Fireworks::execLoop()
{
    cout << "\n\n\tIn execLoop.\n\n";
    SDL_Event e;
    
    quit = false;
    try
    {
        // Setup the window
        if (SDL_Init(SDL_INIT_VIDEO|SDL_INIT_AUDIO) != 0){
            logSDLError(cout, "SDL_Init");
            exit(1);
        }
        else
        {
            cout << "\n\n\tInitialized SDL.\n\n";
        }
        SDL_GL_SetAttribute(SDL_GL_CONTEXT_MAJOR_VERSION, 3);
        SDL_GL_SetAttribute(SDL_GL_CONTEXT_MINOR_VERSION, 0);
        window = SDL_CreateWindow("OpenGL Fireworks Display", 500, 200, SCR_WIDTH, SCR_HEIGHT, SDL_WINDOW_OPENGL|SDL_WINDOW_RESIZABLE);
        if (window == nullptr){
            logSDLError(cout, "CreateWindow");
            SDL_Quit();
            exit(1);
        }
        else
        {
            cout << "\n\n\tCreated SDL window.\n\n";
        }
        
        context = SDL_GL_CreateContext(window);
        if (context == nullptr){
            logSDLError(cout, "CreateContext");
            SDL_DestroyWindow(window);
            SDL_GL_DeleteContext(context);
            SDL_Quit();
            exit(1);
        }
        else
        {
            cout << "\n\n\tCreated SDL context.\n\n";
        }
        renderer = SDL_CreateRenderer(window, -1, SDL_RENDERER_SOFTWARE);
        if (renderer == nullptr){
            logSDLError(cout, "CreateRenderer");
            SDL_DestroyWindow(window);
            SDL_GL_DeleteContext(context);
            SDL_Quit();
            exit(1);
        }
        else
        {
            cout << "\n\n\tCreated renderer.\n";
        }
        SDL_GL_MakeCurrent(window, context);
        glewExperimental=true;
        GLenum err=glewInit();
        if(err!=GLEW_OK)
        {
            //Problem: glewInit failed, something is seriously wrong.
            cout<<"\n\n\tglewInit failed, aborting."<<endl;
            exit(1);
        }
        else
        {
            cout << "\n\n\tInitialized glew.\n";
        }
        glEnable(GL_DEPTH_TEST);  
        glDepthFunc(GL_LESS);
        glDepthRange(0.1f, 1000.0f);
        for (int x = 0; x < NUM_EXPLOSIONS; x++)
        {
            stringstream ss;
            ss << x + 1;
            shader[x] = new Shader();
            shader[x]->initShader("/usr/share/openglresources/shaders/vertex.glsl",
            "/usr/share/openglresources/shaders/fragment.glsl", "shader-" + ss.str() + ".bin");
            cout << "\n\n\tShader1-" << x + 1 << " created.\n\n";
        }
    }
    catch(exception exc)
    {
        cout << "\n\n\tProgram Initialization Error:  " << exc.what() << "\n\n";
    }
    if (debug2)
    {
        cout << "\n\n\tStarting update cycle.\n\n";
    }
    for (int x = 0; x < NUM_EXPLOSIONS; x++)
    {
        update(0, x);
    }
    explData[0].gap = 0;
    thread[0] = SDL_CreateThread((SDL_ThreadFunction)SDL2Fireworks::sndMaker, "TestThread", (void *)NULL);
    explData[0].start = false;
    defineObjects();
    interval = 0;
    // render loop
    // -----------
    while (!quit)
    {
        if (interval > 100000000)
        {
            for (int x = 0; x < NUM_EXPLOSIONS; x++)
            {
                update(0, x);
            }
            explData[0].gap = 0;
            thread[0] = SDL_CreateThread((SDL_ThreadFunction)SDL2Fireworks::sndMaker, "TestThread", (void *)NULL);
            explData[0].start = false;
            interval = 0;
        }
        for (int x = 0; x < NUM_EXPLOSIONS; x++)
        {
            interval += 10;
            if (debug1)
            {
                cout << "\n\n\tProcessing interval:  " << interval << ".\n\n";
            }
            if ((explData[x].gap < (interval - 30)) && (explData[x].start))
            {
                /* Simply create a thread */
                thread[x] = SDL_CreateThread((SDL_ThreadFunction)SDL2Fireworks::sndMaker, "SoundThread", (void *)NULL);

                if (NULL == thread[x]) {
                    logSDLError(cout, "SDL_CreateThread failed");
                } else {
                    if (debug1)
                    {
                        cout << "\n\n\tThread " << x << " successfully created.\n\n";
                    }
                }
                explData[x].start = false;
            }
            if (interval >= explData[x].gap)
            {
                if (interval > (explData[x].gap + 1000))
                {
                    SDL_WaitThread(thread[x], &threadReturnValue);
                    cout << "\n\n\tThread returned value:  " << threadReturnValue << "\n\n";
                    update(interval, x);
                }
                // render
                // ------
                intbegin = chrono::system_clock::now();
                projection = camera->getPerspective();
                view = camera->getViewMatrix();
                glClearColor(0.0f, 0.0f, 0.0f, 1.0f);
                glClearDepthf(1.0f);
                glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
                glViewport(0, 0, width, height);
                shader[x]->Use();
                shader[x]->setFloat("u_time", interval - explData[x].gap);
                shader[x]->setVec4("u_color", explData[x].color);
                shader[x]->setVec4("u_centerPosition", explData[x].centerPos);
                shader[x]->setMat4("view", view);
                shader[x]->setMat4("projection", projection);
                shader[x]->setMat4("scaler", scaler);
                glDrawArrays(GL_POINTS, 0, NUM_PARTICLES);
                intend = chrono::system_clock::now();
                
                while (SDL_PollEvent(&e))
                {
                    keyDown(e);
                    mouseMove(e);
                    windowEvent(e);
                };
                if (debug1)
                {
                    cout << "\n\tSwapping window for:  " << x;
                }
                SDL_GL_SwapWindow(window);
            }
        }
    }
    //! ------------------------------------------------------------------
    SDL_DestroyWindow(window);
    SDL_GL_DeleteContext(context);
    SDL_DestroyRenderer(renderer);
    SDL_Quit();
}

//! ---------------------------------------------------------------------------------------------------------
// process all input: query SDL2 whether relevant keys are pressed/released this frame and react accordingly
// ---------------------------------------------------------------------------------------------------------
void SDL2Fireworks::keyDown(SDL_Event e)
{
    int delta = (int)
    //! Use timing to create a cameraSpeed variable.
    chrono::duration_cast<chrono::nanoseconds>(intend -
    intbegin).count();
    float deltaTime = (float)delta / 1000000.0f;
    float cameraSpeed = 0.5f * deltaTime;
    //! Motion keys.
    //! Forward motion.
    if (e.type == SDL_KEYDOWN) 
    {
        cout << "\n\n\tIn keyDown:  " << SDL_GetKeyName(e.key.keysym.sym) << "\n\n";
        switch (e.key.keysym.sym)
        {
            case SDLK_w:
                camera->processKeyboard(Camera::Camera_Movement::FORWARD, cameraSpeed);
                break;
            //! Backwards motion.
            case SDLK_s:
                camera->processKeyboard(Camera::Camera_Movement::BACKWARD, cameraSpeed);
                break;
            //! Move left.
            case SDLK_a:
                camera->processKeyboard(Camera::Camera_Movement::LEFT, cameraSpeed);
                break;
            //! Move right.
            case SDLK_d:
                camera->processKeyboard(Camera::Camera_Movement::RIGHT, cameraSpeed);
                break;
            //! Move up.
            case SDLK_r:
                camera->processKeyboard(Camera::Camera_Movement::UP, cameraSpeed);
                break;
            //! Move down.
            case SDLK_f:
                camera->processKeyboard(Camera::Camera_Movement::DOWN, cameraSpeed);
                break;
            //! Reset the camera.
            case SDLK_z:
                camera->resetCamera();
                break;
            //! Reverse the camera.
            case SDLK_x:
                camera->reverseDirection();
                break;
            //! Zoom keys.
            //! Zoom in.
            case SDLK_UP:
                camera->processMouseScroll(Camera::Camera_Movement::CLOSER);
                break;
            //! Zoom out.
            case SDLK_DOWN:
                camera->processMouseScroll(Camera::Camera_Movement::AWAY);
                break;
            case SDLK_ESCAPE:
                quit = true;
                break;
           case SDLK_RALT:
                altSet = true;
                break;
            case SDLK_LALT:
                altSet = true;
                break;
            case SDLK_RETURN:
                if (altSet)
                {
                    cout << "\n\n\tSet window fullscreen.\n\n";
                    SDL_SetWindowFullscreen(window, SDL_WINDOW_FULLSCREEN_DESKTOP);
                    altSet = false;
                }
                break;
        }
    }
}    

void SDL2Fireworks::windowEvent(SDL_Event e)
{
    if (e.type == SDL_WINDOWEVENT) {
        switch (e.window.event) {
        case SDL_WINDOWEVENT_SHOWN:
            cout << "\n\n\tWindow " << e.window.windowID << " shown.\n\n";
            break;
        case SDL_WINDOWEVENT_HIDDEN:
            cout << "\n\n\tWindow " << e.window.windowID << " hidden.\n\n";
            break;
        case SDL_WINDOWEVENT_EXPOSED:
            cout << "\n\n\tWindow " << e.window.windowID << " exposed.\n\n";
            break;
        case SDL_WINDOWEVENT_MOVED:
            cout << "\n\n\tWindow " << e.window.windowID << " moved to " 
            << e.window.data1 << ", " << e.window.data2 << "\n\n";
            break;
        case SDL_WINDOWEVENT_RESIZED:
            cout << "\n\n\tWindow " << e.window.windowID << " resized to " 
            << e.window.data1 << ", " << e.window.data2 << "\n\n";
            width = (unsigned  int)e.window.data1;
            height = (unsigned  int)e.window.data2;
            glViewport(0, 0, e.window.data1, e.window.data2);
            camera->resizeView(e.window.data1, e.window.data2);
            defineObjects();
            break;
        case SDL_WINDOWEVENT_SIZE_CHANGED:
            cout << "\n\n\tWindow " << e.window.windowID << " size changed to " 
            << e.window.data1 << ", " << e.window.data2 << "\n\n";
            width = (unsigned  int)e.window.data1;
            height = (unsigned  int)e.window.data2;
            glViewport(0, 0, e.window.data1, e.window.data2);
            camera->resizeView(e.window.data1, e.window.data2);
            defineObjects();
            break;
        case SDL_WINDOWEVENT_MINIMIZED:
            cout << "\n\n\tWindow " << e.window.windowID << " minimized.\n\n";
            width = (unsigned  int)0;
            height = (unsigned  int)0;
            glViewport(0, 0, 0, 0);
            camera->resizeView(0, 0);
            defineObjects();
            break;
        case SDL_WINDOWEVENT_MAXIMIZED:
            cout << "\n\n\tWindow " << e.window.windowID << " maximized.\n\n";
            width = (unsigned int)e.window.data1;
            height = (unsigned int)e.window.data2;
            SDL_SetWindowSize(window, width, height);
            glViewport(0, 0, width, height);
            camera->resizeView(width, height);
            defineObjects();
            break;
        case SDL_WINDOWEVENT_RESTORED:
            cout << "\n\n\tWindow " << e.window.windowID << " restored.\n\n";
            break;
        case SDL_WINDOWEVENT_ENTER:
            cout << "\n\n\tMouse entered window " << e.window.windowID << ".\n\n";
            break;
        case SDL_WINDOWEVENT_LEAVE:
            cout << "\n\n\tMouse left window " << e.window.windowID << ".\n\n";
            break;
        case SDL_WINDOWEVENT_FOCUS_GAINED:
            cout << "\n\n\tWindow " << e.window.windowID << " gained keyboard focus.\n\n";
            break;
        case SDL_WINDOWEVENT_FOCUS_LOST:
            cout << "\n\n\tWindow " << e.window.windowID << " lost keyboard focus.\n\n";
            break;
        case SDL_WINDOWEVENT_CLOSE:
            cout << "\n\n\tWindow " << e.window.windowID << " closed.\n\n";
            quit = true;
            break;
        case SDL_WINDOWEVENT_TAKE_FOCUS:
            cout << "\n\n\tWindow " << e.window.windowID << " is offered focus.\n\n";
            break;
        case SDL_WINDOWEVENT_HIT_TEST:
            cout << "\n\n\tWindow " << e.window.windowID << " has a special hit test.\n\n";
            break;
        default:
            cout << "\n\n\tWindow " << e.window.windowID << " received an unknown event.\n\n";
            break;
        }
    }
}

void SDL2Fireworks::mouseMove(SDL_Event e)
{
    if (e.type == SDL_MOUSEWHEEL)
    {
            if(e.wheel.y > 0)
            {
                camera->processMouseScroll(Camera::Camera_Movement::CLOSER);
            }
            else if (e.wheel.y < 0)
            {
                camera->processMouseScroll(Camera::Camera_Movement::AWAY);
            }
    }
    else if( e.type == SDL_MOUSEMOTION)
    {
        camera->processMouseMovement(e.motion.xrel, e.motion.yrel);
    }
    
}    


void SDL2Fireworks::logSDLError(ostream &os, const string &msg)
{
    os << "\n\n\t" << msg << " error: " << SDL_GetError() << "\n\n";
}

void SDL2Fireworks::defineObjects()
{
    if (VAO > 0)
    {
        cout << "\n\n\tDeleted VAO.\n\n";
        glDeleteBuffers(1, &VAO);
    }
    //Create vertex buffer.
    glGenBuffers(1, &VAO);
    glBindBuffer(GL_ARRAY_BUFFER, VAO);
    glBufferData(GL_ARRAY_BUFFER, sizeof(particles), particles, GL_STATIC_DRAW);
    glVertexAttribPointer(0, 4, GL_FLOAT,
    GL_FALSE, sizeof(ParticleData), (GLvoid*) 0);
    glEnableVertexAttribArray(0);
    glVertexAttribPointer(1, 4, GL_FLOAT,
    GL_FALSE, sizeof(ParticleData), (GLvoid*) (4 * sizeof(float)));
    glEnableVertexAttribArray(1);
    glVertexAttribPointer(2, 4, GL_FLOAT,
    GL_FALSE, sizeof(ParticleData), (GLvoid*) (8 * sizeof(float)));
    glEnableVertexAttribArray(2);
    glBindBuffer(GL_ARRAY_BUFFER, 0);
}    

void SDL2Fireworks::sndMaker()
{
    Uint32 wave_len; // length of our sample
    Uint8 *wave_buf; // buffer containing our audio file
    SDL_AudioSpec wave_spec, ret_spec; // the specs of our piece of music
    if (debug1)
    {
        cout << "\n\n\tStarting sound production.";
        cout << "\n\tLoading WAV file.\n\n";
    }
    
    if (SDL_LoadWAV("/usr/share/openglresources/sounds/explosion.wav", &wave_spec, &wave_buf, &wave_len) ==  NULL)
    {
        cout << "\n\n\tUnable to load WAV file:  " << SDL_GetError() << "\n\n";
        exit(-1);
    }
    else
    {
        if (debug1)
        {
            cout << "\n\n\tWAV file successfully loaded.\n\n";
        }
    }
	// set the callback function
	/*
	wave_spec.callback = SDL2Triangle::my_audio_callback;
	wave_spec.userdata = NULL;
	// set our global static variables
	audio_buf = wave_buf; // copy sound buffer
	audio_len = wave_len; // copy file length
	*/
    string name  = SDL_GetAudioDeviceName(0, 0);
    SDL_AudioDeviceID dev = SDL_OpenAudioDevice(name.c_str(), 0, &wave_spec, &ret_spec, SDL_AUDIO_ALLOW_FORMAT_CHANGE);
    if (dev < 0)
    {
        cout << "\n\n\tUnable to open audio device:  " << SDL_GetError() << "\n\n";
        exit(-1);
    }
    else
    {
        if (debug1)
        {
            if (debug1)
            {
                cout << "\n\n\tAudio device successfully opened.\n\n";
            }
        }    
        int result = SDL_QueueAudio(dev, wave_buf, wave_len);
        if (result != 0)
        {
            cout <<  "\n\n\tUnable to queue audio data.\n\n";
            exit(-1);
        }
        else
        {
            if (debug1)
            {
                cout << "\n\n\tAudio data successfully queued.\n\n";
            }
        }
    }
    SDL_PauseAudioDevice(dev, 0); /* start audio playing. */
    if (debug1)
    {
        cout << "\n\n\tPlayed the audio.\n\n";
    }
    SDL_Delay(800); /* let the audio callback play some sound for 5 seconds. */
    SDL_CloseAudioDevice(dev);
    if (debug1)
    {
        cout << "\n\n\tClosed audio device.\n\n";
    }
	SDL_FreeWAV(wave_buf);
}
